### ISSUE TITLE
Intermittent errors from code coverage mapping (Transformation error, An error occurred while trying to read the map file, ENAMETOOLONG)

### ISSUE THREAD

**Issue by itslenny** [istanbuljs/nyc#847]
## Expected Behavior

Produce coverage report

## Observed Behavior

I was banging my head against a wall trying to get nyc to work with my project then suddenly it was working flawlessly, and I couldn't figure out why. After some messing around I realized that if I modified a file I'd get the error (shown below) for that one file.

I eventually figured out that if I open `package.json` and modified the nyc config and set `"sourceMap": false` and ran the tests they'd run fine (no errors). Then if I set it back to `"sourceMap": true` it would produce coverage as expected.

I've also deleted the `coverage` and `.nyc_output` directories and it doesn't seem to cause the problem to come back. They are re-created as expected when I run `npm test`

### Error Output

```
Transformation error for /src/controllers/root/root.controller.ts ; return original code [removed long base 64 string, presumably source maps]

An error occurred while trying to read the map file at /controllers/root/data:application/json;charset=utf-8;base64,[removed long base 64 string, presumably source maps]

Error: ENAMETOOLONG: name too long, open '/controllers/root/data:application/json;charset=utf-8;base64,[removed long base 64 string, presumably source maps]
```

## Code (or Repository) that Reproduces Issue

The project is a TypeScript Node server using Jasmine for tests.

**npm script (package.json)**

```
"test": "nyc jasmine --config=jasmine.json",
```

**nyc config (package.json)**

```
  "nyc": {
    "extension": [
      ".ts"
    ],
    "exclude": [
      "**/*.spec.ts",
      "**/*.mock.ts",
      "src/test/**/*.*"
    ],
    "reporter": [
      "html"
    ],
    "require": [
      "ts-node/register",
      "source-map-support/register"
    ],
    "sourceMap": true,
    "all": false
  },
```

**jasmine.json**

```
{
  "spec_dir": "./src",
  "spec_files": [
    "**/*.spec.ts"
  ],
  "helpers": [
    "test/test-requirements.ts"
  ],
  "stopSpecOnExpectationFailure": true,
  "random": false
}
```

## Forensic Information

**Operating System:** OSX 10.11.6
**Environment Information:**

Node 9.2.0
nyc 12.0.1
([full details](https://gist.github.com/itslenny/33d5d8d88f9d8f18e63d0fb65b2bd768))

**Comment by itslenny** [comment]
**Update...**

I'd like to find an answer to my problem, but I've realized the intermittent nature of it seems to be a bug in the caching in nyc. If I go into the nyc config and set `"cache": false` the error happens consistently.

Additionally, after looking at the coverage output it doesn't appear to be correctly aligned to the source TypeScript.

My guess is that when toggling source map on and off it's producing coverage without source map and storing that in cache, but not clearing cache when there is an nyc config changes (when I turn source map back on).

**Comment by bcoe** [comment]
@itslenny whenever the issue is intermittent like this, it tends to be the order that the coverage files are loaded and merged from `.nyc_output/`. TypeScript seems to love bumping into issues related to this.

For starters give the `12x` release a try, which swaps out Babel 6 for Babel 7.

**Comment by itslenny** [comment]
@bcoe Thanks for the response.

I was already on 12.0.1

Also, I figured out that if I set `cache:false` in the nyc config it stops being intermittent and is just a consistent error.

**Comment by devonhumes** [comment]
I'm seeing the exact same issue. I'm using jasmine with TypeScript and a very similar setup. Any additional insight into the issue?

**Comment by devonhumes** [comment]
I figured it out for my project. The issue is that there was a double encoding of source maps. Source maps were being generated as external, then they were inlined and finally read as if they were external. I fixed the issue by using the `inlineSourceMap: true` flag in my `tsconfig.json`. I then changed the `nyc` config to:

```
  "nyc": {
    "extension": [
      ".ts"
    ],
    "exclude": [
      "**/*.spec.ts"
    ],
    "reporter": [
      "html"
    ],
    "sourceMap": true,
    "cache": true,
    "require": [
      "ts-node/register"
    ]
  },```

I did test with `cache: false`.

**Comment by bcoe** [comment]
Please see https://github.com/Microsoft/TypeScript/issues/24993, I'm interested in overhauling our TypeScript support and would love your feedback and requirements.

**Comment by devonhumes** [comment]
@niedzielski What are your `tsconfig.json` and `nyc` settings?

**Comment by niedzielski** [comment]
@devonhumes we're using ES6 with Babel, not TypeScript. Here's the patch (with sourceMap disabled): https://gerrit.wikimedia.org/r/#/c/mediawiki/extensions/Popups/+/440904/7/.nycrc.json.

**Comment by itslenny** [comment]
@devonhumes It's working for you? What version(s) are you running?

Here's what I've tried. Let me know if you have any other ideas...

I still can't get it working. I tried adding `"inlineSourceMap": true`, but it didn't change anything. Still getting the "name too long" errors with the base64 sourcemaps in the path.

I tried every combination of `"inlineSourceMap": true` and `"sourceMap": true` (true/false) and never saw a difference.

I also tried including `"source-map-support/register"` in nyc config (require) which I saw in some examples, but it doesn't seem to make a difference with/without it.

The only way it works at all is if I set `sourceMap: false` in nyc config, but then the line numbers are completely wrong in the resulting coverage output.

full nyc config

```
  "nyc": {
    "extension": [
      ".ts"
    ],
    "exclude": [
      "**/*.spec.ts",
      "**/*.mock.ts",
      "src/test/**/*.*"
    ],
    "reporter": [
      "html"
    ],
    "require": [
      "ts-node/register",
      "source-map-support/register"
    ],
    "sourceMap": true,
    "cache": false,
    "all": false
  },
```

**Comment by itslenny** [comment]
Figured it out. I had `"mapRoot": ""` in my tsconfig which is not valid with `"inlineSourceMap": true`. nyc seems to swallow that compiler error and continues to output the "name too long" error.

I found the problem by just running `tsc` directly which told me that I'm not allowed to use `mapRoot` with `inlineSourceMap` and refuses to build. Once I removed `mapRoot` nyc started producing coverage reports as expected.
